packages/kbn-alerts-ui-shared/src/common/hooks/use_virtual_data_view_query.ts

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import type { DataViewsContract } from '@kbn/data-views-plugin/common';
+import { useQuery } from '@tanstack/react-query';
+import { useEffect } from 'react';
+import { QueryOptionsOverrides } from '../types/tanstack_query_utility_types';
+
+export interface UseVirtualDataViewParams {
+  // Dependencies
+  dataViewsService: DataViewsContract;
+
+  // Params
+  /**
+   * The index names to create the data view for
+   */
+  indexNames?: string[];
+}
+
+export const queryKeyPrefix = ['alerts', 'dataView'];
+
+/**
+ * Creates an in-memory data view, cached by index names
+ *
+ * When testing components that depend on this hook, prefer mocking {@link DataViewsContract}'s
+ * create and clearInstanceCache method instead of the hook itself.
+ * @external https://tanstack.com/query/v4/docs/framework/react/guides/testing
+ */
+export const useVirtualDataViewQuery = (
```

Comment:
Here I'm leveraging react-query's async state management and caching even though we're not actually fetching from the network. This allows to avoid risky [`useEffect`s](https://github.com/elastic/kibana/pull/189638/files#diff-8e7de7345f3f13823c76709ff7d5a53f65664d2a12dcea8fa686291282457b4fL92), make sure the data view is created only once per query key and encapsulates the teardown logic